portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,19 +14,88 @@
 
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.sps.data.Comment;
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.List;
+import com.google.gson.Gson;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  //private ArrayList<String> arr = new ArrayList<String>(); 
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello Andrew!</h1>");
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    // populate array with data from the DB
+    List<Comment> comments = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+        long id = entity.getKey().getId();
+        String name = (String) entity.getProperty("name");
+        String comment_content = (String) entity.getProperty("comment");
+        long timestamp = (long) entity.getProperty("timestamp");
+
+        Comment comment = new Comment(id, name, comment_content, timestamp);
+        comments.add(comment);
+    }
+
+    // Convert arr to JSON
+    String json = convertToJson(comments); 
+    
+    // Send response to client
+    response.setContentType("application/json");
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+      // Get data from request
+      String name = request.getParameter("comment_name");
+      String comment =  request.getParameter("comment_content");
+
+      // Add to Datastore
+      Entity commentEntity = new Entity("Comment");
+      commentEntity.setProperty("name", name);
+      commentEntity.setProperty("comment", comment);
+      commentEntity.setProperty("timestamp", System.currentTimeMillis());
+
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      datastore.put(commentEntity);
+
+      // pause the thread to ensure data is in DB before page refresh
+      // could not find equivalent of async call to datastore (tried transactions
+      // but couldn't get them to work)
+      try {
+        Thread.sleep(10);
```

Comment:
Yikes!  I'm glad your instinct was to try another way, but I don't think I'd check this in even if it seems to work.  10 milliseconds might be enough right now, for you, on this cluster, at this time of day and time of the week, but there's no guarantee it will be for everyone all the time.

Unfortunately I don't know much about this API, but it's surprising to me that `put()` can return before the data is actually present.  If that's true, I would hope it's documented, which should suggest what you should do to deal with it.  I'm happy to help if you can point me at the relevant docs!